### PR TITLE
fix(Core/Util): EventMap::RepeatEvent when eventId is retrieved by ExecuteEvent() call

### PR DESCRIPTION
**Pull request by sanctum32**
## CHANGES PROPOSED:
-  When eventMap.ExecuteEvent() is called, it has chance to loose current eventId since this function can erase triggered event from its map. This is major issue if developers tries to use BossAI scriptable virtual function "void ExecuteEvent(uint32 eventId)" or eventMap.ExecuteEvent() call

- Originally this idea is taken from TrinityCore, base idea author of these changes is @Trisjdc (https://github.com/TrinityCore/TrinityCore/commit/8db02c52acbdbf2c3d0280a01ca6a8b367eb5617)

-  Code optimization and cleanup

- Removed PopEvent and GetEvent functions

## TESTS PERFORMED:
- Tested with still in progress rewrite for AhnKahet (https://github.com/sanctum32/azerothcore-wotlk, ahnkahet_revamp branch)

- Wotlk 5 man dungeons, ICC, Ruby Sanctum, Ulduar bosses, npcs

## HOW TO TEST THE CHANGES:

- This PR basically effects every dungeon, content, raid, basically TBC, WOTLK raids, dungeons needs minimal testing, take attention if unit should do repeatable actions (in particular phases or timers, not meaning special actions)

-  For programmers: this can be tested by invoking eventMap.ExecuteEvent or by BossAI scriptable virtual function "ExecuteEvent"

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by sanctum32** [comment]
Noting, this base idea was taken from TrinityCore project, i'm not sure who made original idea/fix in their core

**Comment by FrancescoBorzi** [comment]
@sanctum32  about:

> This can be tested by invoking eventMap.ExecuteEvent or by BossAI scriptable virtual function "ExecuteEvent"

there is no need to do that, we just need a way to test the game and make sure that this change doesn't break anything existing.

Keep in mind that testers aren't necessarily programmers, so we can't ask testers to invoke methods etc... Instead: check where this method is already invoked and suggest to testers to give that a try in game.

**Comment by sanctum32** [comment]
well, currently most of code which uses ExecuteEvent
triggers stright schedulle again
seems none in core there wasnt any attempt to combine EventRepeat together with ExecuteEvent
but this PR does exactly what title is saying
this mostly makes EventRepeat possible together with EventExecute
advantages of this is simple
if there is phased or grouped events
during repeat calls you do not need to repeat phase, group, eventId indexes byself

File: src/common/Utilities/Util.h [hunk 1]
```diff
@@ -765,7 +761,10 @@ class EventMap
                 else if (_phase && (itr->second & 0xFF000000) && !(itr->second & (_phase << 24)))
                     _eventMap.erase(itr);
```
Comment (Kitzunu) [inline src/common/Utilities/Util.h:762]: It'd be kind if you could enclose this in brackets too :D

**Comment by sanctum32** [comment]
This PR can be tested with 5 man dungeons, like azjol nerub or ahnkahet

File: src/common/Utilities/Util.h (hunk 1 above)
Comment (sanctum32) [reply src/common/Utilities/Util.h:762]: its not part of my PR code but fine

**Comment by sanctum32** [comment]
since last event id was preserved inside of EventMap, i retired two functions which are no longer needed, it means, C++ developers wont need to manually stop event after having GetEvent() calls, also this will perfectly will work with RepeatEvent calls, none of events will get lost

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @pklloveyou @geekena @locus313  please give this one a try

**Comment by Winfidonarleyan** [comment]
Well well well, Is it really? This PR easy testing, need testing bosses 2+ phases if the phase changes, then everything works.

**Comment by locus313** [comment]
tested, played through a couple multi phase bosses. No issues found so far.  Will run a few more tests over the next couple days.

**Comment by Winfidonarleyan** [comment]
I have been using changes #2660 for 2 months, I have problems with 2+ phases in all bosses, and I've solved that problem with this commit https://github.com/TrinityCore/TrinityCore/commit/8db02c52acbdbf2c3d0280a01ca6a8b367eb5617

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by sanctum32** [comment]
Alright thanks, i will include its author into this PR description, not meaning to overtake overal credits ;) i mostly need this PR for my own rewrite which i'm doing for AhnKahet (ofcourse rest dungeons will follow). Thanks for testing

**Comment by locus313** [comment]
Did some more testing, played through another couple multi phase bosses. No issues found.

**Comment by sanctum32** [comment]
I maybe will make file movement to external source file later. I think changes shall be visible in this PR for rest developers to see what actually was changed because if I will move this implementation to external file, file changes won't be easyly visible, just I will make that in other PR

**Comment by FrancescoBorzi** [comment]
@sanctum32 there are conflicts

File: src/common/Utilities/Util.h [hunk 2]
```diff
@@ -713,54 +713,44 @@ class EventMap
...
+    * @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
+    */
+    void RepeatEvent(uint32 time)
+    {
+        _eventMap.insert(EventStore::value_type(_time + time, _lastEvent));
+    }
+
+        if (itr->first > _time)
+            return 0;
+        else if (_phase && (itr->second & 0xFF000000) && !((itr->second >> 24) & _phase))
```
Comment (Helias) [inline src/common/Utilities/Util.h:731]: add the curly braces here pls
moreover, you could use a constant for `0xFF000000`, I saw that it is used more times.

File: src/common/Utilities/Util.h [hunk 3]
```diff
@@ -772,171 +762,181 @@ class EventMap
...
+                delayed.insert(EventStore::value_type(itr->first + delay, itr->second));
+                _eventMap.erase(itr++);
             }
             else
+            {
                 ++itr;
+            }
         }
+        else
+            ++itr;
```
Comment (Helias) [inline src/common/Utilities/Util.h:827]: Should be better merge all `if`s conditions into one and put only one else with `++itr;` ?

File: src/common/Utilities/Util.h [hunk 4]
```diff
@@ -772,171 +762,181 @@ class EventMap
...
+                itr = _eventMap.begin();
             }
             else
+            {
                 ++itr;
+            }
         }
+        else
+            ++itr;
     }
```
Comment (Helias) [inline src/common/Utilities/Util.h:889]: same here

**Review by Helias — CHANGES_REQUESTED** [review]

File: src/common/Utilities/Util.h (hunk 2 above)
Comment (sanctum32) [reply src/common/Utilities/Util.h]: its kinda... no offense but this file got effected by astyle commit which triggered mass of conflicts, when conflict was resolved, NOW this gets into inspection for style? Seriously why didn't used Astyle to add auto brackets...

Comment (sanctum32) [reply src/common/Utilities/Util.h]: Resolved

**Comment by FrancescoBorzi** [comment]
@Si1ker can you please test this?

**Review by Helias — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @iyasenai can you please have a look at this?

**Comment by Knindzagxg** [comment]
I just had 1 small issue, but everything else was good.

After killing 3 ritual adds, Svala stand in the air quite a while.
Not sure does it have to do something with this PR or it's script by herself.
It should go down after 5 sec once last (3rd) add is killed.
WoWScrnShot_101520_172158

**Comment by FrancescoBorzi** [comment]
I doubt that this PR was the cause. @sanctum32 can you please confirm?

**Comment by FrancescoBorzi** [comment]
from @sanctum32 on Discord

> my pr mostly eliminates events.PopEvent();, at any way Execute just automates that

**Pull request merged**